### PR TITLE
Issue #8973 - Rework KeyStoreScanner handling for symlink related changes

### LINKED ISSUE #8973: Does ssl-reload monitor keystore when it is a symlink?
**jetty-9.4.49.v20220914** and **11.0.12**

**Java 11**

**Does ssl-reload monitor keystore when it is a symlink?**

Have deployed my application which uses jetty in kubernetes pod. We are using cert-manger for certificate management.
Secret is mounted inside a pod as below:

```
drwxr-xr-x    2 root     root         140 Nov 29 09:38 ..2022_11_29_09_38_19.297914115
lrwxrwxrwx    1 root     root          31 Nov 29 09:38 ..data -> ..2022_11_29_09_38_19.297914115
lrwxrwxrwx    1 root     root          13 Nov 29 07:06 ca.crt -> ..data/ca.crt
lrwxrwxrwx    1 root     root          19 Nov 29 07:06 keystore.p12 -> ..data/keystore.p12
lrwxrwxrwx    1 root     root          14 Nov 29 07:06 tls.crt -> ..data/tls.crt
lrwxrwxrwx    1 root     root          14 Nov 29 07:06 tls.key -> ..data/tls.key
lrwxrwxrwx    1 root     root          21 Nov 29 07:06 truststore.p12 -> ..data/truststore.p12

```

When the certificates are renewed automatically by cert-manager , the target file for symlink keystore.p12 is changed , but even after that jetty is not picking up the updated certs. If keystore is not symlink , jetty is picking the updated certs.

### PR DESCRIPTION
**Pull request by joakime**
Fixes #8973

+ Removed changes from #8786 and #8787
+ More test cases

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
@joakime you need to update the `jetty-ssl-context-reload.xml` and `ssl-reload.mod` files as well.

**Comment by joakime** [comment]
@lachlan-roberts done, please re-review

**Comment by lachlan-roberts** [comment]
@joakime I made a commit to remove some bad documentation about the followLinks setting from `module-ssl-reload.adoc`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java [hunk 1]
```diff
@@ -88,11 +78,23 @@ public KeyStoreScanner(SslContextFactory sslContextFactory, boolean followLinks)
...
+            return keystoreFile.toPath();
+        }
+    }
+
     @Override
     public void fileAdded(String filename)
     {
         if (LOG.isDebugEnabled())
-            LOG.debug("added {}", filename);
+            LOG.debug("fileAdded {} - keystoreFile.toReal {}", filename, getRealKeyStorePath());
```
Comment (lachlan-roberts) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java:97]: I don't think we need to do a `toRealPath()` for debugging when we don't do it for the monitored file or for the scanner.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java [hunk 2]
```diff
@@ -88,11 +78,23 @@ public KeyStoreScanner(SslContextFactory sslContextFactory, boolean followLinks)
         addBean(_scanner);
     }
 
+    private Path getRealKeyStorePath()
+    {
+        try
+        {
+            return keystoreFile.toPath().toRealPath();
```
Comment (lachlan-roberts) [inline jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java:85]: You could assign `keystoreFile.toPath()` to a variable and prevent doing this twice in the case of a failure.
(not that this matters that much as its only for debugging)

**Review by lachlan-roberts — APPROVED** [review]
I can't think of an example where it would be necessary to resolve the symlinks in the `KeyStoreScanner`, so I think removing `followLinks` setting is ok.

File: jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/ssl/KeyStoreScanner.java:97]: it helps, in debugging, to see that it picked up the right file.

**Pull request merged**